staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature_support_checker.go

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/apiserver/pkg/storage"
+)
+
+var (
+	// This is a error type that we used in isDeprecated
+	ErrNoRequestWatchProgressSupport = errors.New("")
+	// Define these static versions to use for checking version of etcd, issue on kubernetes #123192
+	v3_4_25 = version.MustParseSemantic("3.4.25")
+	v3_5_0  = version.MustParseSemantic("3.5.0")
+	v3_5_8  = version.MustParseSemantic("3.5.8")
+)
+
+// FeatureSupportChecker to define Supports function which checks etcd with assigned feature and return boolean.
+type FeatureSupportChecker interface {
+	Supports(ctx context.Context, feature storage.Feature) (bool, error)
+}
+
+// featureSupportInfo holds information about feature support status and associated error.
+type featureSupportInfo struct {
+	Supported bool
+	Err       error
+}
+
+type DefaultFeatureSupportChecker struct {
+	client *clientv3.Client
+	// mutext protects featureSupportCache variable.
+	mutex sync.RWMutex
+	// featureSupportCache store result of if feature is supported or not.
+	featureSupportCache map[storage.Feature]featureSupportInfo
+}
+
+func newDefaultFeatureSupportChecker(client *clientv3.Client) *DefaultFeatureSupportChecker {
+	return &DefaultFeatureSupportChecker{
+		client:              client,
+		featureSupportCache: make(map[storage.Feature]featureSupportInfo),
+	}
+}
+
+// Supports checks whether a specific feature is supported by etcd.
+// TODO: Extend this implementation to support feature-checking for the cacher.
+func (c *DefaultFeatureSupportChecker) Supports(ctx context.Context, feature storage.Feature) (bool, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// Check if the feature is already cached
+	if info, ok := c.featureSupportCache[feature]; ok {
+		return info.Supported, info.Err
+	}
+
+	// Feature-specific support check
+	supported, err := c.checkFeatureSupport(ctx, feature)
+
+	// Cache the result
+	c.featureSupportCache[feature] = featureSupportInfo{
```

Comment:
This is missing some kind of retries.

If my apiserver started earlier than etcd itself, we clearly won't be able to check the version so it will end-up being an error - but it's not something we want to keep forever.

Reply:
Right, we should distinguish between errors not being up from others.

Reply:
I put retries that only triggers on `Status` returning an error please take a look. @wojtek-t @serathius